apps/recovery-relay/lib/wallets/TON/index.ts

```diff
@@ -0,0 +1,106 @@
+import { Ton as BaseTon } from '@fireblocks/wallet-derivation';
+import { ConnectedWallet } from '../ConnectedWallet';
+import { TonClient, WalletContractV4 } from '@ton/ton';
+import { beginCell, Cell, fromNano } from '@ton/core';
+import { AccountData } from '../types';
+import { defaultTonWalletV4R2code } from './tonParams';
+import axios from 'axios';
+
+export class Ton extends BaseTon implements ConnectedWallet {
+  public rpcURL: string | undefined;
+  public setRPCUrl(url: string): void {
+    this.rpcURL = url;
+  }
+  private client = new TonClient({
+    endpoint: this.isTestnet ? 'https://testnet.toncenter.com/api/v2/jsonRPC' : 'https://toncenter.com/api/v2/jsonRPC',
+  });
+  private tonWallet = WalletContractV4.create({ publicKey: Buffer.from(this.publicKey.replace('0x', ''), 'hex'), workchain: 0 });
+
+  public async getBalance(): Promise<number> {
+    await new Promise((resolve) => setTimeout(resolve, 2000));
+    const contract = this.client.open(this.tonWallet);
+    return Number(fromNano(await contract.getBalance()));
+  }
+  public async broadcastTx(tx: string): Promise<string> {
+    try {
+      const body = Cell.fromBoc(Buffer.from(tx, 'base64'))[0];
+      const pubKey = Buffer.from(this.publicKey.replace('0x', ''), 'hex');
+
+      const externalMessage = beginCell().storeUint(0b10, 2).storeUint(0, 2).storeAddress(this.tonWallet.address).storeCoins(0);
+
+      const seqno = await this.getSeqno();
+      if (seqno === 0) {
+        // for the fist transaction we initialize a state init struct which consists of init struct and code
+        externalMessage
+          .storeBit(1) // We have State Init
+          .storeBit(1) // We store State Init as a reference
+          .storeRef(this.createStateInit(pubKey)); // Store State Init as a reference
+      } else {
+        externalMessage.storeBit(0); // We don't have state init
+      }
+      const finalExternalMessage = externalMessage.storeBit(1).storeRef(body).endCell();
+
+      await new Promise((resolve) => setTimeout(resolve, 2000));
+      await this.client.sendFile(finalExternalMessage.toBoc());
+      const txHash = finalExternalMessage.hash().toString('hex');
+      this.relayLogger.debug(`TON: Tx broadcasted: ${txHash}`);
+      return txHash;
+    } catch (e) {
+      this.relayLogger.error(`TON: Error broadcasting tx: ${e}`);
+      if (axios.isAxiosError(e)) {
+        this.relayLogger.error(`Axios error: ${e.message}\n${e.response?.data}`);
+      }
```

Comment:
Add the error code for ton so in case of an issue we can direct to the ton documentation

Reply:
You mean the error code returned from the node? It is part of the response.data I print already if that's the case